File: 백준/Gold/10026. 적록색약/적록색약.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

from collections import deque
dx, dy = (1, -1, 0, 0), (0, 0, 1, -1)

def main():
    n = int(input())
    board = [list(input()) for _ in range(n)]
    visited = [[0]*n for _ in range(n)]
    cnt1 = 0
    for i in range(n):
        for j in range(n):
            if not visited[i][j]:
                c = board[i][j]
                dq = deque([(i,j)])
                visited[i][j] = 1
                while dq:
                    x, y = dq.popleft()
                    for dir in range(4):
                        nx, ny = x+dx[dir], y+dy[dir]
                        if not (0 <= nx < n) or not (0 <= ny < n): continue
                        if board[nx][ny] == c and not visited[nx][ny]:
                            visited[nx][ny] = 1
                            dq.append((nx, ny))
                cnt1 += 1
    for i in range(n):
        for j in range(n):
            if board[i][j] == 'G': board[i][j] = 'R'
    cnt2 = 0
    visited = [[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if not visited[i][j]:
                c = board[i][j]
                dq = deque([(i,j)])
                visited[i][j] = 1
                while dq:
                    x, y = dq.popleft()
                    for dir in range(4):
                        nx, ny = x+dx[dir], y+dy[dir]
                        if not (0 <= nx < n) or not (0 <= ny < n): continue
                        if board[nx][ny] == c and not visited[nx][ny]:
                            visited[nx][ny] = 1
                            dq.append((nx, ny))
                cnt2 += 1
    print(cnt1, cnt2)
if __name__ == '__main__':
    main()

File: 백준/Gold/10026. 적록색약/적록색약.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pii = pair<int, int>;

const int dx[] = {1, -1, 0, 0}, dy[] = {0, 0, 1, -1};

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n; cin >> n;
    vector<vector<char>> board(n, vector<char>(n));
    for(int i = 0; i < n; i++) for(int j = 0; j < n; j++) cin >> board[i][j];
    vector<vector<bool>> visited(n, vector<bool>(n, 0));
    int cnt1, cnt2; cnt1 = cnt2 = 0;
    for(int i = 0; i < n; i++) for(int j = 0; j < n; j++){
        if(!visited[i][j]){
            char c = board[i][j];
            queue<pii> dq; dq.push({i, j});
            visited[i][j] = 1;
            while(dq.size()){
                auto [x, y] = dq.front(); dq.pop();
                for(int dir = 0; dir < 4; dir++){
                    int nx = x+dx[dir], ny = y+dy[dir];
                    if(0 > nx || nx >= n || 0 > ny || ny >=n) continue;
                    if(board[nx][ny] == c && !visited[nx][ny]){
                        visited[nx][ny] = 1;
                        dq.push({nx, ny});
                    }
                }
            }
            cnt1++;
        }
    }
    for(int i = 0; i < n; i++) for(int j = 0; j < n; j++){
        if(board[i][j] == 'G') board[i][j] = 'R';
        visited[i][j] = 0;
    }
    for(int i = 0; i < n; i++) for(int j = 0; j < n; j++){
        if(!visited[i][j]){
            char c = board[i][j];
            queue<pii> dq; dq.push({i, j});
            visited[i][j] = 1;
            while(dq.size()){
                auto [x, y] = dq.front(); dq.pop();
                for(int dir = 0; dir < 4; dir++){
                    int nx = x+dx[dir], ny = y+dy[dir];
                    if(0 > nx || nx >= n || 0 > ny || ny >=n) continue;
                    if(board[nx][ny] == c && !visited[nx][ny]){
                        visited[nx][ny] = 1;
                        dq.push({nx, ny});
                    }
                }
            }
            cnt2++;
        }
    }
    cout << cnt1 << ' ' << cnt2;
    return 0;
}